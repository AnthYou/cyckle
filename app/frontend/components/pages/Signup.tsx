import React, { useState } from "react";
import { Link } from "react-router-dom";

import Button from "../UI/Button";
import ManDoor from "@/images/login.svg";
import Input from "../UI/Input";
import Select, { SelectOptions } from "../UI/Select";

interface SignUpFormTypes {
  email: string;
  password: string;
  passwordConfirmation: string;
  firstName: string;
  lastName: string;
  gender: string;
  height: number;
  phone: string;
}

const Signup = () => {
  const defaultValues: SignUpFormTypes = {
    email: "",
    password: "",
    passwordConfirmation: "",
    firstName: "",
    lastName: "",
    gender: "male",
    height: 150,
    phone: "",
  };

  const genderOptions: SelectOptions[] = [
    { id: 1, label: "Homme", value: "male" },
    { id: 2, label: "Femme", value: "female" },
    { id: 3, label: "Non-binaire", value: "non-binary" },
  ];

  const [formValues, setFormValues] = useState<SignUpFormTypes>(defaultValues);

  const isPasswordValid =
    formValues.password === formValues.passwordConfirmation &&
    formValues.password.length >= 6;
  const areMinCharsValid =
    formValues.firstName.length >= 3 &&
    formValues.lastName.length >= 2 &&
    formValues.phone.length >= 10;
  const isHeightValid = formValues.height >= 100 && formValues.height <= 200;
  const isValid = isPasswordValid && areMinCharsValid && isHeightValid;

  const handleInputChange = (
    e:
      | React.ChangeEvent<HTMLInputElement>
      | React.ChangeEvent<HTMLSelectElement>
  ) => {
    const { name, value } = e.target;

    setFormValues((prevValues) => {
      return {
        ...prevValues,
        [name]: value,
      };
    });
  };

  const handleForm = (event: React.FormEvent) => {
    event.preventDefault();

    console.log(formValues);
  };

  return (
    <>
      <h1>Inscrivez-vous</h1>
      <div className="flex flex-col sm:flex-row items-center justify-between gap-4">
        <form onSubmit={handleForm} className="w-full sm:w-1/2">
          <Input
            id="firstName"
            type="text"
            label="Prénom"
            name="firstName"
            placeholder="Raymond"
            value={formValues.firstName}
            onChange={handleInputChange}
            required={true}
          />
          <Input
            id="lastName"
            type="text"
            label="Nom de famille"
            name="lastName"
            placeholder="Poulidor"
            value={formValues.lastName}
            onChange={handleInputChange}
            required={true}
          />
          <Select
            id="gender"
            name="gender"
            label="Genre"
            options={genderOptions}
            value={formValues.gender}
            onChange={handleInputChange}
            required={true}
          />
          <Input
            id="height"
            type="number"
            label="Taille (en cm)"
            name="height"
            min="100"
            max="200"
            value={formValues.height}
            onChange={handleInputChange}
            required={true}
          />
          <Input
            id="email"
            type="text"
            label="Email"
            name="email"
            value={formValues.email}
            onChange={handleInputChange}
            placeholder="raymond@cyckle.com"
            required={true}
          />
          <Input
            id="phone"
            type="tel"
            label="Numéro de téléphone"
            name="phone"
            value={formValues.phone}
            onChange={handleInputChange}
            placeholder="06 01 02 03 04"
            required={true}
          />
          <Input
            id="password"
            type="password"
            label="Mot de passe"
            name="password"
            placeholder="******"
            value={formValues.password}
            onChange={handleInputChange}
            required={true}
          />
          <Input
            id="passwordConfirmation"
            type="password"
            label="Confirmation du mot de passe"
            name="passwordConfirmation"
            placeholder="******"
            value={formValues.passwordConfirmation}
            onChange={handleInputChange}
            required={true}
          />
          <div className="flex flex-col items-center justify-center gap-4">
            <Button color="primary" className="mt-5" disabled={!isValid}>
              S'inscrire
            </Button>
            <p>
              Vous avez déjà un compte ? <Link to="#">Identifiez-vous</Link>
            </p>
          </div>
        </form>
        <img
          src={ManDoor}
          alt="A man opening a door."
          className="w-full sm:w-1/2"
        />
      </div>
    </>
  );
};

export default Signup;
