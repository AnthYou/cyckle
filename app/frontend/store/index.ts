import { Action, configureStore, ThunkAction } from "@reduxjs/toolkit";
import authReducer from "./slices/auth-slice";

const store = configureStore({
  reducer: { auth: authReducer }
});

export default store;
export type RootState = ReturnType<typeof store.getState>;
export type AppDispatch = typeof store.dispatch;
export type AppThunk<ReturnType = void> = ThunkAction<
  ReturnType,
  RootState,
  unknown,
  Action<string>
>;
