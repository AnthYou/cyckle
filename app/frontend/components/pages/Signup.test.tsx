import React from "react";
import { Provider } from "react-redux";
import { BrowserRouter } from "react-router-dom";
import renderer from "react-test-renderer";

import store from "@/store/index";
import Signup from "./Signup";

describe("<Signup />", () => {
  it("should render correctly", () => {
    const component = renderer.create(
      <Provider store={store}>
        <BrowserRouter>
          <Signup />
        </BrowserRouter>
      </Provider>
    );

    const tree = component.toJSON();

    expect(tree).toMatchSnapshot();
  });
});
