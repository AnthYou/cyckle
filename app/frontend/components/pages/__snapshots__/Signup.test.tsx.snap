// Vitest Snapshot v1

exports[`<Signup /> > should render correctly 1`] = `
[
  <h1>
    Inscrivez-vous
  </h1>,
  <div
    className="flex flex-col sm:flex-row items-center justify-between gap-4"
  >
    <form
      className="w-full sm:w-1/2"
      onSubmit={[Function]}
    >
      <div
        className="_Input_658f0a"
      >
        <label
          htmlFor="firstName"
        >
          Prénom
        </label>
        <input
          id="firstName"
          name="firstName"
          onChange={[Function]}
          placeholder="Raymond"
          required={true}
          type="text"
          value=""
        />
      </div>
      <div
        className="_Input_658f0a"
      >
        <label
          htmlFor="lastName"
        >
          Nom de famille
        </label>
        <input
          id="lastName"
          name="lastName"
          onChange={[Function]}
          placeholder="Poulidor"
          required={true}
          type="text"
          value=""
        />
      </div>
      <div
        className="_Select_dd8b4d"
      >
        <label
          htmlFor="gender"
        >
          Genre
        </label>
        <select
          id="gender"
          name="gender"
          onChange={[Function]}
          required={true}
          value="male"
        >
          <option
            value="male"
          >
            Homme
          </option>
          <option
            value="female"
          >
            Femme
          </option>
          <option
            value="non-binary"
          >
            Non-binaire
          </option>
        </select>
      </div>
      <div
        className="_Input_658f0a"
      >
        <label
          htmlFor="height"
        >
          Taille (en cm)
        </label>
        <input
          id="height"
          name="height"
          onChange={[Function]}
          placeholder="180"
          required={true}
          type="number"
          value={150}
        />
      </div>
      <div
        className="_Input_658f0a"
      >
        <label
          htmlFor="email"
        >
          Email
        </label>
        <input
          id="email"
          name="email"
          onChange={[Function]}
          placeholder="raymond@cyckle.com"
          required={true}
          type="text"
          value=""
        />
      </div>
      <div
        className="_Input_658f0a"
      >
        <label
          htmlFor="phone"
        >
          Numéro de téléphone
        </label>
        <input
          id="phone"
          name="phone"
          onChange={[Function]}
          placeholder="06 01 02 03 04"
          required={true}
          type="tel"
          value=""
        />
      </div>
      <div
        className="_Input_658f0a"
      >
        <label
          htmlFor="password"
        >
          Mot de passe
        </label>
        <input
          id="password"
          name="password"
          onChange={[Function]}
          placeholder="******"
          required={true}
          type="password"
          value=""
        />
      </div>
      <div
        className="_Input_658f0a"
      >
        <label
          htmlFor="passwordConfirmation"
        >
          Confirmation du mot de passe
        </label>
        <input
          id="passwordConfirmation"
          name="passwordConfirmation"
          onChange={[Function]}
          placeholder="******"
          required={true}
          type="password"
          value=""
        />
      </div>
      <div
        className="flex flex-col items-center justify-center gap-4"
      >
        <button
          className="_Button_4ea4ec _primary_4ea4ec mt-5"
        >
          S'inscrire
        </button>
        <p>
          Vous avez déjà un compte ? 
          <a
            href="/"
            onClick={[Function]}
          >
            Identifiez-vous
          </a>
        </p>
      </div>
    </form>
    <img
      alt="A man opening a door."
      className="w-full sm:w-1/2"
      src="/@fs/@/images/login.svg"
    />
  </div>,
]
`;
